### PR TITLE
Optimize verification and implement batch verification for TurboPlonk

### PR DESCRIPTION
**Pull request by confuseSUN**
* **The major changes of this PR**

* **The major impacts of this PR**
  - [ ] Impact WASM?
  - [ ] Impact mainnet data compatibility?

* **Extra documentations**

### CODE REVIEW

**Comment by weikengchen** [comment]
@confuseSUN We probably still need to put all the methods that use "parallel" i.e. Rayon behind a feature.

The reason is mostly for WASM. If one compiles with WASM, we may explicitly disable parallelization since the Rayon compatibility with WASM is just so-so now.

**Comment by confuseSUN** [comment]
> 

okay, I will change it.

File: api/benches/anon_xfr.rs [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+};
+use zei_accumulators::merkle_tree::{PersistentMerkleTree, Proof, TreePath};
+use zei_algebra::{bls12_381::BLSScalar, prelude::*};
+use zei_crypto::basic::{pedersen_comm::PedersenCommitmentRistretto, rescue::RescueInstance};
+
+const AMOUNT: u64 = 10u64;
+const ASSET: AssetType = AssetType([1u8; ASSET_TYPE_LENGTH]);
+const BATCHSIZE: [usize; 7] = [1, 2, 3, 6, 10, 20, 30];
+
+// Measurement of the verification time and batch verification tim of `abar_to_abar`.
```
Comment (findora-crypto) [inline api/benches/anon_xfr.rs:41]:
```suggestion
// Measurement of the verification time and batch verification time of `abar_to_abar`.
```

File: api/benches/anon_xfr.rs [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+fn bench_abar_to_abar(c: &mut Criterion) {
+    let outputs = vec![(5, FEE_TYPE), (15, FEE_TYPE), (30, ASSET)];
+    let inputs = vec![(20 + 13 as u64, FEE_TYPE), (30, ASSET)];
+    abar_to_abar(c, inputs, outputs, 13);
+}
+
+// Measurement of the verification time and batch verification tim of `abar_to_bar`.
```
Comment (findora-crypto) [inline api/benches/anon_xfr.rs:48]:
```suggestion
// Measurement of the verification time and batch verification time of `abar_to_bar`.
```

File: api/benches/anon_xfr.rs [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+fn bench_abar_to_bar(c: &mut Criterion) {
+    abar_to_bar(c);
+}
+
+// Measurement of the verification time and batch verification tim of `abar_to_ar`.
```
Comment (findora-crypto) [inline api/benches/anon_xfr.rs:53]:
```suggestion
// Measurement of the verification time and batch verification time of `abar_to_ar`.
```

File: api/benches/anon_xfr.rs [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+fn bench_abar_to_ar(c: &mut Criterion) {
+    abar_to_ar(c);
+}
+
+// Measurement of the verification time and batch verification tim of `bar_to_abar`.
```
Comment (findora-crypto) [inline api/benches/anon_xfr.rs:58]:
```suggestion
// Measurement of the verification time and batch verification time of `bar_to_abar`.
```

File: api/benches/anon_xfr.rs [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+fn bench_bar_to_abar(c: &mut Criterion) {
+    bar_to_abar(c);
+}
+
+// Measurement of the verification time and batch verification tim of `ar_to_abar`.
```
Comment (findora-crypto) [inline api/benches/anon_xfr.rs:63]:
```suggestion
// Measurement of the verification time and batch verification time of `ar_to_abar`.
```

File: api/src/anon_xfr/abar_to_abar.rs [hunk 6]
```diff
@@ -232,6 +234,68 @@ pub fn verify_anon_xfr_note<D: Digest<OutputSize = U64> + Default>(
     .c(d!(ZeiError::AXfrVerificationError))
 }
 
+/// Batch verify the anonymous transfer notes.
```
Comment (findora-crypto) [inline api/src/anon_xfr/abar_to_abar.rs:237]:
```suggestion
/// Batch verify the anonymous transfer notes.
/// Note: this function assumes that the correctness of the Merkle roots has been checked outside.
```

File: api/src/anon_xfr/abar_to_ar.rs [hunk 7]
```diff
@@ -209,6 +211,73 @@ pub fn verify_abar_to_ar_note<D: Digest<OutputSize = U64> + Default>(
     .c(d!(ZeiError::AXfrVerificationError))
 }
 
+/// Batch verify the anonymous-to-transparent notes.
```
Comment (findora-crypto) [inline api/src/anon_xfr/abar_to_ar.rs:214]:
```suggestion
/// Batch verify the anonymous-to-transparent notes.
/// Note: this function assumes that the correctness of the Merkle roots has been checked outside.
```

File: api/src/anon_xfr/abar_to_bar.rs [hunk 8]
```diff
@@ -352,6 +354,146 @@ pub fn verify_abar_to_bar_note<D: Digest<OutputSize = U64> + Default>(
     .c(d!(ZeiError::AXfrVerificationError))
 }
 
+/// Batch verify the anonymous-to-confidential notes.
```
Comment (findora-crypto) [inline api/src/anon_xfr/abar_to_bar.rs:357]:
```suggestion
/// Batch verify the anonymous-to-confidential notes.
/// Note: this function assumes that the correctness of the Merkle roots has been checked outside.
```

**Pull request merged**